mod-clans/src/main/java/gg/solarmc/loader/clans/ClanManager.java

```diff
@@ -0,0 +1,141 @@
+     * @throws IllegalStateException if the clan isn't present in table
+     */
+    public Clan getClan(Transaction transaction, Integer id) {
+        return clans.get(id, i -> {
+
+            var jooq = transaction.getProperty(DSLContext.class);
+
+            ClansClanInfoRecord rec = jooq.fetchOne(CLANS_CLAN_INFO,CLANS_CLAN_INFO.CLAN_ID.eq(i));
+
+            if (rec == null) {
+                throw new IllegalStateException("No such clan exists!");
+            }
+
+            Set<ClanMember> members = jooq.select(CLANS_CLAN_MEMBERSHIP.USER_ID).from(CLANS_CLAN_MEMBERSHIP)
+                    .where(CLANS_CLAN_MEMBERSHIP.CLAN_ID.eq(id)).fetchSet((rec1) -> new ClanMember(i,rec1.value1(),this));
+
+            ClansClanAlliancesRecord rec1 = jooq.fetchOne(CLANS_CLAN_ALLIANCES,CLANS_CLAN_ALLIANCES.CLAN_ID.eq(i));
+
+            ClanMember owner = new ClanMember(i,rec.getClanLeader(),this);
+
+            //Yes, i know i could :probably: put rec1#getAllyId in the alliedClan slot and save 2 lines of code
+            //However i don't want to risk any bugs and nullpointerexceptions
+            //i hate null you were right a248
+
+            if (rec1 == null) {
+                return new Clan(rec.getClanId(), rec.getClanName(),rec.getClanKills(),
+                        rec.getClanDeaths(),rec.getClanAssists(),null,this,members,owner);
+            } else {
+                return new Clan(rec.getClanId(), rec.getClanName(),rec.getClanKills(),
+                        rec.getClanDeaths(),rec.getClanAssists(),rec1.getAllyId(),this,members,owner);
+            }
+
+        });
+    }
+
+    /**
+     * Creates an empty clan with given name.
+     * @param name Name of the gg.solarmc.loader.clans.Clan to add
+     * @param transaction the tx
+     * @param owner the to be owner of the clan
+     * @return created clan.
+     */
+    public Clan createClan(Transaction transaction, String name, ClanDataObject owner) {
+        ClansClanInfoRecord rec = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_INFO)
+                .columns(CLANS_CLAN_INFO.CLAN_NAME,CLANS_CLAN_INFO.CLAN_LEADER,CLANS_CLAN_INFO.CLAN_KILLS,CLANS_CLAN_INFO.CLAN_DEATHS,CLANS_CLAN_INFO.CLAN_ASSISTS)
+                .values(name,owner.getUserId(),0,0,0)
+                .returning()
+                .fetchOne();
+
+        if (rec == null) throw new IllegalStateException("Failed to insert new gg.solarmc.loader.clans.Clan by name " + name);
+
+        ClanMember ownerAsMember = owner.asClanMember(transaction);
+        Set<ClanMember> memberSet = new HashSet<>();
+        memberSet.add(ownerAsMember);
+
+        Clan returned = new Clan(rec.getClanId(),rec.getClanName(),rec.getClanKills(),rec.getClanDeaths(),rec.getClanAssists(),null,this,memberSet,ownerAsMember);
+
+        clans.put(returned.getID(),returned);
+
+        return returned;
+    }
+
+    public void deleteClan(Transaction transaction, Clan clan) {
```

Comment:
Needs javadoc. The javadoc might seem annoying but really there's always something it answers. For example, what happens if the clan does not exist? Whether nothing happens, or an exception is thrown, is important information

Reply:
Fixed